consensus/integration_test_types.go

```diff
@@ -0,0 +1,443 @@
+//
+//	B1 - B2 - B3
+//
+// such that B1 is the parent of B2, and B2 is an ancestor of B3. The ancestor-descendant relationship is established
+// whenever a block contains the QC for another block. We say that this block is the descendant of the other block.
+// In particular, if the two blocks were proposed with consecutive views, we say these blocks are in a parent-child
+// relationship. So, when we observe the aforementioned configuration of B1, B2, and B3, we finalize all ancestors of
+// B1 as well as B1. To see why this is safe, one is referred to read the Fast-HotStuff paper.
+func (node *validatorNode) commitChainFromGrandParent(block *block) {
+	// In accordance to the above comment, B3 = block, B2 = parent, and B1 = grandParent.
+	parent := node.safeBlocks[block.GetQC().GetBlockHash().GetValue()]
+	if parent == nil {
+		return
+	}
+
+	// We verify that B2 is the parent of B3.
+	if block.GetView() != (parent.GetView() + 1) {
+		return
+	}
+
+	grandParent := node.safeBlocks[parent.GetQC().GetBlockHash().GetValue()]
+	if grandParent == nil {
+		return
+	}
+
+	// We verify that B1 is the parent of B2.
+	if parent.GetView() != (grandParent.GetView() + 1) {
+		return
+	}
+
+	// We have successfully observed a committing configuration, we will now commit all ancestors of B1 as well as B1.
+	node.latestCommittedBlock = grandParent
+}
+
+func (node *validatorNode) ProcessVote(vote VoteMessage) {
+	node.lock.Lock()
+	defer node.lock.Unlock()
+
+	if node.eventLoop.status != eventLoopStatusRunning {
+		return
+	}
+
+	node.eventLoop.ProcessValidatorVote(vote)
+}
+
+func (node *validatorNode) ProcessTimeout(timeout TimeoutMessage) {
+	node.lock.Lock()
+	defer node.lock.Unlock()
+
+	if node.eventLoop.status != eventLoopStatusRunning {
+		return
+	}
+
+	node.eventLoop.ProcessValidatorTimeout(timeout)
+}
+
+func (node *validatorNode) Start() {
+	node.eventLoop.Start()
+	go node.runEventSignalLoop()
+}
+
+func (node *validatorNode) runEventSignalLoop() {
+	for {
+		select {
+		case event := <-node.eventLoop.Events:
+			switch event.EventType {
+			case FastHotStuffEventTypeVote:
+				node.handleVoteEvent(event)
+				break
+			case FastHotStuffEventTypeTimeout:
+				node.handleTimeoutEvent(event)
+				break
+			case FastHotStuffEventTypeConstructVoteQC:
+				node.handleVoteQCConstructionEvent(event)
+				break
+			case FastHotStuffEventTypeConstructTimeoutQC:
+				node.handleTimeoutQCConstructionEvent(event)
+				break
+			}
+		case <-node.quit:
+			return
+		}
+	}
+}
+
+func (node *validatorNode) handleVoteEvent(event *FastHotStuffEvent) {
+	payload := GetVoteSignaturePayload(event.View, event.TipBlockHash)
+	signature, err := node.privateKey.Sign(payload[:])
+	if err != nil {
+		panic(err)
+	}
+
+	vote := &voteMessage{
+		view:      event.View,
+		blockHash: event.TipBlockHash,
+		publicKey: node.privateKey.PublicKey(),
+		signature: signature,
+	}
+
+	for _, validator := range node.validatorNodes {
+		go validator.ProcessVote(vote)
```

Comment:
I think a goroutine might be OK here because there's no way that the view could increment as a result of processing a vote. But, to be safe, I would call ProcessVote synchronously for the current node. And I would do the same for all the other functions.

Reply:
Same comment as above. This is safe because all handle.*() functions hold a lock on the event loop when processing incoming messages / events.